### DISCUSSION TITLE
Provide RateLimit info

### DISCUSSION THREAD

**Discussion by rogerallen** [MarshalX/atproto#167] [Ideas]
Hi there, I'm currently working on debugging why my bot hits rate limit issues and I don't know how to access this info via atproto.  

I see logging.debug messages on certain calls will print out this info:
```
DEBUG:httpcore.http11:receive_response_headers.complete return_value=(
  b'HTTP/1.1', 200, b'OK', 
  [(b'Date', b'Sat, 23 Sep 2023 23:02:24 GMT'), 
   (b'Content-Type', b'application/json; charset=utf-8'), 
   ... 
   (b'RateLimit-Limit', b'30'), 
   (b'RateLimit-Remaining', b'26'), 
   (b'RateLimit-Reset', b'1695510188'), 
   (b'RateLimit-Policy', b'30;w=300'), ...
  ])
```
But I don't know how to query this `RateLimit` info via the atproto library.

**Comment by MarshalX** [comment]
Hi. Typically when you are rate-limited exception will be raised. The exception provides access to the XrpcError instance and response headers. In headers, you can find information about limits. I guess there is no way to receive headers without reaching the limit. 

Exception example:
![image](https://github.com/MarshalX/atproto/assets/15520314/1d52e1bf-d9ab-48c1-b773-0c59cd9ae86d)

**Comment by rogerallen** [comment]
Seems like it would be best to proactively limit yourself if you are getting close to rate limits. 

Could/should I make this an Request For Enhancement (RFE)?

**Reply by MarshalX** [reply]
It will be awesome to share your vision with the public API of how SDK should provide it. My first thoughts about new methods in the
Client to get rate limits headers from the latest request 

Upd. Just access to response headers. Without strict data models because it changes. It is not described in lexicons and so on.

**Reply by rogerallen** [reply]
I think I agree, access to the response headers of latest reply might be most the most reasonable, but this really is not my area of expertise.  

It seems unfortunate that the protocol doesn't list the name of the current limit that is being applied.  If that were the case, keeping a dictionary indexed by name of the limit could be possible.  As is, multiple limit names could have the same policy, so policy cannot be the key in that dictionary.

[Happily, updating to the latest version of atproto seems to have resolved the ratelimit issues I had been having]

**Comment by MarshalX** [comment]
Hello. I prepared simple example of override method that gives you access to response headers (which contains rate limits).

```python
class RateLimitedClient(Client):
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

        self._limit = self._remaining = self._reset = None

    def get_rate_limit(self):
        return self._limit, self._remaining, self._reset

    def _invoke(self, *args, **kwargs):
        response = super()._invoke(*args, **kwargs)

        self._limit = response.headers.get('ratelimit-limit')
        self._remaining = response.headers.get('ratelimit-remaining')
        self._reset = response.headers.get('ratelimit-reset')

        return response
```

Usage:
```python
client = RateLimitedClient()
client.login(os.environ['USERNAME'], os.environ['PASSWORD'])

post = client.get_timeline()
print(client.get_rate_limit())
```

**Reply by ariok** [reply]
> Hello. I prepared simple example of override method that gives you access to response headers (which contains rate limits).
> 
> ```python
> class RateLimitedClient(Client):
>     def __init__(self, *args, **kwargs) -> None:
>         super().__init__(*args, **kwargs)
> 
>         self._limit = self._remaining = self._reset = None
> 
>     def get_rate_limit(self):
>         return self._limit, self._remaining, self._reset
> 
>     def _invoke(self, *args, **kwargs):
>         response = super()._invoke(*args, **kwargs)
> 
>         self._limit = response.headers.get('RateLimit-Limit')
>         self._remaining = response.headers.get('RateLimit-Remaining')
>         self._reset = response.headers.get('RateLimit-Reset')
> 
>         return response
> ```
> 
> Usage:
> 
> ```python
> client = RateLimitedClient()
> client.login(os.environ['USERNAME'], os.environ['PASSWORD'])
> 
> post = client.get_timeline()
> print(client.get_rate_limit())
> ```

This code should be part of the default Client class.

**Reply by MarshalX** [reply]
@ariok headers are not part of the protocol. And they could be different from instance to instance. It will work with Bluesky, but with another service no. That's why it is still not in the SDK. Access to response headers could be improved; I agree.

**Reply by ariok** [reply]
> @ariok headers are not part of the protocol. And they could be different from instance to instance. It will work with Bluesky, but with another service no. That's why it is still not in the SDK. Access to response headers could be improved; I agree.

I see your point. 

Just a quick fix in case anyone else is reading this. In my case (on bluesky) to have your code working I had to lowercase the headers properties.

```
        self._limit = response.headers.get('ratelimit-limit')
        self._remaining = response.headers.get('ratelimit-remaining')
        self._reset = response.headers.get('ratelimit-reset')
```

 Thank you again for your work on this repo!

**Reply by loloster** [reply]
as @MarshalX told me i have modifed the new class as:

```
class RateLimitedClient(Client):
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

        self._limit = self._remaining = self._reset = None

    def get_rate_limit(self):
        return self._limit, self._remaining, self._reset

    def _invoke(self, *args, **kwargs):
        response = super()._invoke(*args, **kwargs)

        #print all headers at invocation
        print(response.headers)
        self._limit = response.headers.get('ratelimit-limit')
        self._remaining = response.headers.get('ratelimit-remaining')
        self._reset = response.headers.get('ratelimit-reset')

        return response
```
i was experiencing ratelimit at the login phase so in my code i added :

```
     client.login('****.bsky.social', '****-****-****-****')
     print(client.get_rate_limit())
```
here what i get :
```
{'Date': 'Tue, 22 Oct 2024 20:35:10 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'X-Powered-By': 'Express', 'Access-Control-Allow-Origin': '*', 'RateLimit-Limit': '100', 'RateLimit-Remaining': '18', 'RateLimit-Reset': '1729705510', 'RateLimit-Policy': '100;w=86400', 'ETag': 'W/"5e1-zRw8Z0fLy8hjYLScfNNdbs0BSWk"', 'Vary': 'Accept-Encoding', 'Content-Encoding': 'gzip'}
{'x-powered-by': 'Express', 'access-control-allow-origin': '*', 'cache-control': 'private', 'vary': 'Authorization, Accept-Encoding', 'ratelimit-limit': '3000', 'ratelimit-remaining': '2999', 'ratelimit-reset': '1729629610', 'ratelimit-policy': '3000;w=300', 'content-type': 'application/json; charset=utf-8', 'atproto-repo-rev': '3l74uo22kxj2u', 'atproto-content-labelers': 'did:plc:ar7c4by46qjdydhdevvrndac;redact', 'content-length': '820', 'etag': 'W/"334-atn3oBdlSwHi5RI2a/GX7r781ZA"', 'date': 'Tue, 22 Oct 2024 20:35:10 GMT', 'keep-alive': 'timeout=90', 'strict-transport-security': 'max-age=63072000'}
('3000', '2999', '1729629610')
```
note the different case letters of the headers, a mix of capital and lower case letters on the first block, lowercase only on the second. in oder to get any result thing instead of "None", as @ariok said, one have to ask for the lowercase headers, but what he gets are the values of the mix case letters headers

weird ?

**Reply by loloster** [reply]
i should add that the values of the ratelimit experienced are the one that are present in the first block.
values which are not accessible.
the only consistant lowercase header is ratelimit-reset
the two oher, RateLimit-Limit and RateLimit-Remaining are not accessible and those two are the one we are experiencing
hope that is more clear
sorry

**Reply by loloster** [reply]
in my code the only place where i get the two headers block (and wrong values experienced) are at the login phase
below some snippets of  code and the reponse associated printed on the console
except for the login phase the other seems ok (did not try to hit the limits myself)

HTH

code:
```
client = RateLimitedClient()
print(client.get_rate_limit())
```
response:
`(None, None, None)`
code:
```
client.login('****.bsky.social', '****-****-****-****')
print(client.get_rate_limit())
```
response:
```
{'Date': 'Tue, 22 Oct 2024 21:01:30 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'X-Powered-By': 'Express', 'Access-Control-Allow-Origin': '*', 'RateLimit-Limit': '100', 'RateLimit-Remaining': '17', 'RateLimit-Reset': '1729705510', 'RateLimit-Policy': '100;w=86400', 'ETag': 'W/"5e1-6pYCYYd+AUVctQcYj//Cjy0Q+Ts"', 'Vary': 'Accept-Encoding', 'Content-Encoding': 'gzip'}
{'x-powered-by': 'Express', 'access-control-allow-origin': '*', 'cache-control': 'private', 'vary': 'Authorization, Accept-Encoding', 'ratelimit-limit': '3000', 'ratelimit-remaining': '2999', 'ratelimit-reset': '1729631191', 'ratelimit-policy': '3000;w=300', 'content-type': 'application/json; charset=utf-8', 'atproto-repo-rev': '3l74uugdtfp2v', 'atproto-content-labelers': 'did:plc:ar7c4by46qjdydhdevvrndac;redact', 'content-length': '820', 'etag': 'W/"334-ecVnukYHJaOGEbWkm9BH/JXQ2ds"', 'date': 'Tue, 22 Oct 2024 21:01:31 GMT', 'keep-alive': 'timeout=90', 'strict-transport-security': 'max-age=63072000'}
('3000', '2999', '1729631191')
```
code:
```
thumb_blob = client.upload_blob(img_data).blob
print(client.get_rate_limit())
```
response:
```
{'x-powered-by': 'Express', 'access-control-allow-origin': '*', 'cache-control': 'private', 'vary': 'Authorization, Accept-Encoding', 'ratelimit-limit': '1000', 'ratelimit-remaining': '832', 'ratelimit-reset': '1729638751', 'ratelimit-policy': '1000;w=86400', 'content-type': 'application/json; charset=utf-8', 'content-length': '138', 'etag': 'W/"8a-OQEjAPz7H48movVYbfkUaXdJMSE"', 'date': 'Tue, 22 Oct 2024 21:01:34 GMT', 'keep-alive': 'timeout=90', 'strict-transport-security': 'max-age=63072000'}
thumb_blob= mime_type='image/png' size=4226 ref=IpldLink(link='bafkreicum547zebok7cuguv2qhk4jl62ba26ilofbqeb4sschjs6yo4px4') py_type='blob'
('1000', '832', '1729638751')
```
code:
```
client.send_post(text=text, facets=facets, embed=embed_external)
print(client.get_rate_limit())
```
response:
```
{'x-powered-by': 'Express', 'access-control-allow-origin': '*', 'cache-control': 'private', 'vary': 'Authorization, Accept-Encoding', 'ratelimit-limit': '5000', 'ratelimit-remaining': '4997', 'ratelimit-reset': '1729634494', 'ratelimit-policy': '5000;w=3600', 'content-type': 'application/json; charset=utf-8', 'content-length': '276', 'etag': 'W/"114-i5TPyeybPWJ4PacXjWiUw1F9FG4"', 'date': 'Tue, 22 Oct 2024 21:01:34 GMT', 'keep-alive': 'timeout=90', 'strict-transport-security': 'max-age=63072000'}
('5000', '4997', '1729634494')
```

**Reply by loloster** [reply]
Should have read more doc, just discover session login : no more too short login RateLimit inaccessible value.
```
     if os.path.isfile("session.txt"):
      print('Try to use session.txt to log in')
      with open('session.txt') as f:
       session_string = f.read()
       CLogin=client.login(session_string=session_string)
     else:
      CLogin=client.login('****.atproto.rocks', ****-****-****-****')
     session_string = client.export_session_string()
     with open('session.txt', 'w') as f:
      f.write(session_string)
```

**Reply by MarshalX** [reply]
@loloster thank you! The different casing of header names is fixed in https://github.com/MarshalX/atproto/pull/415. Now they always lowercased. Also I edited my original code snippet to use lowercased keys
